v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/utils/BigtableUtils.java

```diff
@@ -0,0 +1,305 @@
+        BigtableUtils.bigtableRowColumnFamilyName,
+        ChangelogColumns.LOW_WATERMARK.getColumnNameAsByteBuffer(),
+        entry.getTimestamp(),
+        getByteBufferFromString(String.valueOf(entry.getLowWatermark()), charset)
+    ));
+
+    if (entry.getColumn() != null) {
+      // column
+      cells.add(new com.google.cloud.teleport.bigtable.BigtableCell(
+          BigtableUtils.bigtableRowColumnFamilyName,
+          ChangelogColumns.LOW_WATERMARK.getColumnNameAsByteBuffer(),
+          entry.getTimestamp(),
+          getByteBufferFromString(String.valueOf(entry.getLowWatermark()), charset)
+      ));
+    }
+
+    if (entry.getTimestamp() != null) {
+      // timestamp
+      cells.add(new com.google.cloud.teleport.bigtable.BigtableCell(
+          BigtableUtils.bigtableRowColumnFamilyName,
+          ChangelogColumns.TIMESTAMP.getColumnNameAsByteBuffer(),
+          entry.getTimestamp(),
+          getByteBufferFromString(String.valueOf(entry.getTimestamp()), charset)
+      ));
+    }
+
+    if (entry.getTimestampFrom() != null) {
+      // timestamp_from
+      cells.add(new com.google.cloud.teleport.bigtable.BigtableCell(
+          BigtableUtils.bigtableRowColumnFamilyName,
+          ChangelogColumns.TIMESTAMP_FROM.getColumnNameAsByteBuffer(),
+          entry.getTimestamp(),
+          getByteBufferFromString(String.valueOf(entry.getTimestampFrom()), charset)
+      ));
+    }
+
+    if (entry.getTimestampTo() != null) {
+      // timestamp_to
+      cells.add(new com.google.cloud.teleport.bigtable.BigtableCell(
+          BigtableUtils.bigtableRowColumnFamilyName,
+          ChangelogColumns.TIMESTAMP_TO.getColumnNameAsByteBuffer(),
+          entry.getTimestamp(),
+          getByteBufferFromString(String.valueOf(entry.getTimestampTo()), charset)
+      ));
+    }
+
+    if (entry.getValue() != null) {
+      // value
+      cells.add(new com.google.cloud.teleport.bigtable.BigtableCell(
+          BigtableUtils.bigtableRowColumnFamilyName,
+          ChangelogColumns.TIMESTAMP_FROM.getColumnNameAsByteBuffer(),
+          entry.getTimestamp(),
+          getByteBufferFromString(String.valueOf(entry.getValue()), charset)
+      ));
+    }
+
+    return new BigtableRow(
+        createChangelogRowKey(entry.getCommitTimestamp(), workerId, counter, charset),
+        cells
+    );
+  }
+
+  private static ByteBuffer getByteBufferFromString(String s, Charset charset) {
+    return ByteBuffer.wrap(s.getBytes(charset));
+  }
+
+  private static ByteBuffer createChangelogRowKey(
+      Long commitTimestamp,
+      String workerId,
+      AtomicLong counter,
+      Charset charset) {
+    String rowKey = (commitTimestamp.toString()
+        + BigtableUtils.bigtableRowKeyDelimiter
+        + workerId
+        + BigtableUtils.bigtableRowKeyDelimiter
+        + counter.incrementAndGet());
+
+    return ByteBuffer.wrap(rowKey.getBytes(charset));
+  }
+
+  public static List<ChangelogEntry> getValidEntries(
```

Comment:
I think the usefulness of ChangelogEntry class and getValidEntries() is diminished by the fact that other use-cases such as CDC->BQ will have slightly different schema: https://screenshot.googleplex.com/BhKfB8gAnN3Bmva also some stuff can be using BQ type (TIMESTAMP) but optionally can be made numeric instead. 

I think what would would better is having ChangelogEntry housed in CDC->GCS package and making BigtableUtils be strictly agnostic of anything specific to CDC->GCS, which includes not using ChangelogEntry class. For instance, it could just expose something to filter entries based on column name and column family, but avoid generating the entire data object.

Alternatively, which I think it simpler and more preferable, make BigtableUtils a CDC->GCS specific and house it in CDC->GCS package.

Reply:
Agreed. Will be moving this to the bigtablechangestreamstogcs package.